bundles/core/src/main/java/com/adobe/cq/wcm/core/components/commons/link/Link.java

```diff
@@ -0,0 +1,61 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2021 Adobe
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.commons.link;
+
+import java.util.Map;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.osgi.annotation.versioning.ConsumerType;
+
+import com.day.cq.wcm.api.Page;
+
+/**
+ * Describes a link target.
+ */
+@ConsumerType
+public interface Link {
+
+    /**
+     * Check if the link defined for the component is valid.
+     * @return true if component has a valid link defined
+     */
+    boolean isValid();
+
+    /**
+     * Externalized link URL.
+     * @return Link URL or null if link is invalid.
+     */
+    @Nullable
+    String getURL();
+
+    /**
+     * Map with Attributes for HTML Anchor tag for this links.
+     * This usually also contains the Link URL as <code>href</code> attribute,
+     * but may contain additional attributes like <code>target</code> and others.
+     * @return Map with HTML-specific anchor attributes, or an empty map if link is invalid
+     */
+    @NotNull
+    Map<String, String> getHtmlAttributes();
+
+    /**
+     * Returns the references target page if the links points to an internal page.
+     * @return Target page or null
+     */
+    @Nullable
+    Page getTargetPage();
```

Comment:
My personal opinion: I wouldn't even allow getting the target page for a link. 
If models want to provide pages, they can provide pages. I feel that including a mechanism to go backwards from a link to a page is going to lead people to abuse the Link interface as a vehicle for passing Page object references. 

This would be fine, except for the fact that the concept of a Link should be (IMHO) target-agnostic (e.g. page, asset, external, whatever); thus, it's a weak interface that can't be trusted to provide the target page of a link and should not be relied upon to provide that Page object. It requires me to know how the implementation that provided the Link constructed it.

It should be just as valid to add a `getTargetAsset()`, or half a dozen other things that it could point to. 

If we really wanted a reliable mechanism for doing this then I think we would need to have a `PageLink` and `AssetLink` extensions of `Link`, that promise to provide the underlying [not null] object.

Reply:
i think it's important to have access to target page (and yes, also target asset) - think of providing additional metadata for client data layer for example. but we can make it a bit less obvious to make clear it should not be used just to build a link.
we could make the Link interface adaptable, and allow to adapt to a CQ Page or asset, if the link point to any of those.

Reply:
I agree with @ky940819 on this one, having something like `PageLink` and `AssetLink` that extend `Link` make for a better contract.

Reply:
Having thought a bit about this, I think it's a good use-case for generics, as proposed in d7ed7a2.

Then, `Teaser` could define `Link<Page> getLink()` while `Download` could define `Link<Asset> getLink()` and `Link#getReference` will return the correct type. It will also be easy to add new types of references without needing to add extra interfaces. WDYT @stefanseifert & @ky940819 ?

Reply:
generics are eliminated at run time, and if you get a link from a model/link handler it's difficult to detect if it's link to a page or asset - you would have to get the object returned be getTarget and test it's type. 
it makes usage of the Link interface it a bit inconvenient as in most case you just deal with the link and do not care about the targeted type, so you always have to use Link<?> instead of just Link (or get probably warnings in IDE that you do not specify the generics type).

Reply:
Indeed, generics are removed at run time but Java is statically-typed so all checks happen at initial compilation, if all the handlers and models respect the generics annotations, there will be no need for checking of type at runtime.

So yes, the helpers and models would need to specify the generics in the return type of their `getLink`. But I believe they would always know what that link should point to.

A handler that can work with multiple target types will probably expose different methods: `Link<Page> getPageLink(...)` and `Link<Asset> getAssetLink(...)`. It would need to have different logic for them anyway.

Reply:
@stefanseifert , unless there's a clear -1 from you, I would like to merge these changes to the `development` branch, along with [Stage 2](#1413). We will allow for a couple more weeks until including the changes in a release (we need at least to prepare all feature branches for [Stage 3]), that will give us some time to revisit and adjust the interfaces.

Reply:
agreed - this PR gets a bit confusing as it mixes up the fundamentals of API discussion for link handling (which we already did on the first PR in 2019 or so) with the normal code review. it's fine to have new views and opinions on the best API design, but it get's cluttered up in this PR. probably it would have been better to put up a PR with only the interfaces first and dicusss only the public APIs - it's the only thing we cannot change easily after the first release.

Reply:
PR for the interface should have been [Stage 0], my bad 🤦 ! I'll merge this as is and continue the refactoring of `LinkHandler` implementation in [Stage 2] PR (#1413).